### ISSUE TITLE
VerneMQ vs eMQTT

### ISSUE THREAD

**Issue by drasko** [erlio/vernemq#83]
Hello,
I'd like to compare VerneMQ benchmarks with these given here: http://emqtt.io/ for eMQTT (which really seem impressive).

Where can I find this data, and can someone of you VerneMQ expert give us a quick comparioson and benefits of using one or other.

BR,
Drasko

**Comment by dergraf** [comment]
Hi Drasko

we have no official benchmarks published. However, others on the web did so e.g. http://acro-engineer.hatenablog.com/entry/2015/12/09/120500 (it's in Japanese though). We also tweeted a while back some graphs of an internal benchmark, which you can find here. https://twitter.com/vernemq/status/651802735801729024.

The main differences are the way VerneMQ implements clustering and subscriber/message-storage. An additional difference is, interesting for some users, that VerneMQ is an official product developed and pushed by the Swiss company Erl.io.

Hope this helped. Please get in touch with us over email, so we can share with you some details.

Best,
Andre

**Comment by unbalancedparentheses** [comment]
@dergraf it would be awesome to have some basic public benchmarks available. I have a spare 32GB dedicated server if you need some place where to do some load testing.

**Comment by ashleyconnor** [comment]
How does LevelDB in vernemq differ from emqtt's use of mnesia?

Are offline messages distributed to nodes or just in a single place?

**Comment by ioolkos** [comment]
Offline messages are currently stored on one node only, yes.
I'm afraid I can't describe the differences between Mnesia and LevelDB quickly... it requires reading up on the internals and topics like consistency/eventual consistency :)

**Comment by ioolkos** [comment]
@unbalancedparentheses just seen your note...thanks a lot for your kind offer. lots of stuff to do, but we  will also work a little on testing tools. might take you up on that!

**Comment by drasko** [comment]
@dergraf what is the clustering difference between eMQTT and VerneMQ you are mentioning?

**Comment by dergraf** [comment]
Please excuse this wall of text, but it is a complex topic, and I am sure I can link to this response at some other time again. :) I think a short video could also explain some of the ideas here too.

### TL;DR
The main difference relies on the underlying distributed storage engine. Emqtt and RabbitMQ use Mnesia (http://erlang.org/doc/apps/mnesia/Mnesia_overview.html) for its underlying distributed storage engine. VerneMQ uses an eventual consistent datastore backed by LevelDB. 
We decided to **not** use Mnesia because Mnesia doesn't tolerate network partitions very well. Before starting VerneMQ our company consulted on RabbitMQ setups and we experienced those issues hands-on. This made us following a completely different approach to enable clustering.

### Differences
Emqtt (as well as RabbitMQ) relies on Mnesia, as the underlying distributed storage engine (subscriptions, configurations, and retained messages). Mnesia is a distributed database that is packaged together with the Erlang programming language, so it might be the logical choice for this kind of data when you implement a message broker. In fact VerneMQ used it in the beginning too. While using Mnesia is super practical at first, we soon discovered that we require a different approach for replicating this kind of data. We decided to use the Plumtree implementation found in Basho's great Riak distributed KV store and adapted it to use LevelDB.

Basically Plumtree is an epidemic broadcast protocol published in http://repositorio.ul.pt/bitstream/10451/14105/1/07-14.pdf by J Leitao, J Pereira, L Rodrigues from the University Lisbon. The paper describes an algorithm to efficiently broadcast messages to thousands of nodes in a always changing network. Neither Riak nor VerneMQ clusters contain thousands of nodes, but the characteristics of such networks exist in todays computing environments where cluster nodes are dynamically removed or added , cluster nodes may die, network partitions occur, and network latencies are real. (see https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing)

Basho's implementation of the Plumtree protocol for Erlang clusters also added a basic distributed storage engine, which Riak is using for distributing cluster metadata in an eventual consistent manner. VerneMQ currently uses it to distribute subscriber and retain data. So we named it, this kind of storage is **eventual consistent**, with all its pros and cons (read here more about the eventual consistency model https://en.wikipedia.org/wiki/Eventual_consistency). Mnesia on the other hand is built for ACID transactions, of course it also supports dirty read and write operations (giving up most if not all consistency guarantees in a clustered setup), which is used by most performance critical applications (including emqtt). Most of the points why we moved away from Mnesia are summarized in the blog post "Mnesia and CAP" found here https://medium.com/@jlouis666/mnesia-and-cap-d2673a92850#.cpt6ul1qp (it's from 2014 but most if not all points are still valid).

At the point when we decided to switching gears (from Mnesia to Plumtree) the interesting blog post "Call me maybe rabbitmq" found here https://aphyr.com/posts/315-call-me-maybe-rabbitmq was published, which demonstrated several downsides of RabbitMQ and its underlying clustering abilities (mainly it's behaviour during network partitions), which are mostly related to the way RabbitMQ is using the Mnesia database. I am not sure how exactly Emqtt is using Mnesia, but chances are high it's using it in a similar way as RabbitMQ does. After the mentioned blog post RabbitMQ further clarified it's netsplit behaviour found here https://www.rabbitmq.com/partitions.html. The first sentence puts it quite well: _RabbitMQ clusters do not tolerate network partitions well._

### To sum up
All in all we think it's great to have a MQTT broker (with all the great benefits of Erlang) that uses a different approach to distributed storage. However, in times where the network is operational and the cluster state is as expected you won't feel this as the main differentiator between Emqtt and VerneMQ. But it might give you that warm fuzzy feeling.
In case you're interested we have a dedicated chapter on network partitions in our docs https://vernemq.com/docs/clustering/netsplits.html. If you want to dig deeper, and you're familiar with Erlang, you may find the common test suite interesting, where we test different netsplit scenarios.
https://github.com/erlio/vmq_server/blob/master/test/vmq_cluster_netsplit_SUITE.erl

**Comment by drasko** [comment]
@dergraf wow, thanks a lot for such a detailed answer. It definitely deserves a blog post, I think.

Very useful.
